test/integration/certificates/signer_plugin_test.go

```diff
@@ -0,0 +1,143 @@
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	clientset "k8s.io/client-go/kubernetes"
+	restclient "k8s.io/client-go/rest"
+
+	"k8s.io/kubernetes/test/integration/framework"
+)
+
+// Verifies that the CSR approval admission plugin correctly enforces that a
+// user has permission to sign CSRs for the named signer
+func TestCSRSignerNameSigningPlugin(t *testing.T) {
+	_, s, closeFn := framework.RunAMaster(nil)
+	defer closeFn()
+
+	client := clientset.NewForConfigOrDie(&restclient.Config{Host: s.URL, ContentConfig: restclient.ContentConfig{GroupVersion: &schema.GroupVersion{Group: "", Version: "v1"}}})
+
+	csr1 := buildTestingCSR("csr-1", "example.com/signer-1")
+	csr2 := buildTestingCSR("csr-2", "example.com/signer-2")
+
+	testUsername := "testuser"
+	// grant 'testuser' permission to approve CSRs for the signerName 'signer-1' ONLY.
+	cleanupFn, err := grantUserPermissionToSignFor(client, testUsername, "example.com/signer-1")
+	if err != nil {
+		t.Errorf("unable to create test fixture RBAC rules: %v", err)
+		return
+	}
+	defer cleanupFn()
+
+	// create test fixtures
+	if csr1, err = client.CertificatesV1beta1().CertificateSigningRequests().Create(csr1); err != nil {
+		t.Errorf("unable to create test csr: %v", err)
+		return
+	}
+	defer client.CertificatesV1beta1().CertificateSigningRequests().Delete(csr1.Name, nil)
+	if csr2, err = client.CertificatesV1beta1().CertificateSigningRequests().Create(csr2); err != nil {
+		t.Errorf("unable to create test csr: %v", err)
+		return
+	}
+	defer client.CertificatesV1beta1().CertificateSigningRequests().Delete(csr2.Name, nil)
+
+	// approve both CSRs using admin credentials
+	csr1.Status.Conditions = append(csr1.Status.Conditions, certv1beta1.CertificateSigningRequestCondition{
+		Type:    certv1beta1.CertificateApproved,
+		Reason:  "AutoApproved",
+		Message: "Approved during integration test",
+	})
+	if csr1, err = client.CertificatesV1beta1().CertificateSigningRequests().UpdateApproval(csr1); err != nil {
+		t.Fatalf("failed to approve test fixture: %v", err)
+	}
+
+	csr2.Status.Conditions = append(csr2.Status.Conditions, certv1beta1.CertificateSigningRequestCondition{
+		Type:    certv1beta1.CertificateApproved,
+		Reason:  "AutoApproved",
+		Message: "Approved during integration test",
+	})
+	if csr2, err = client.CertificatesV1beta1().CertificateSigningRequests().UpdateApproval(csr2); err != nil {
+		t.Fatalf("failed to approve test fixture: %v", err)
+	}
+
+	// testuserClient is a clientset that impersonates 'testUsername', used for
+	// testing RBAC rules
+	testuserClient := clientset.NewForConfigOrDie(&restclient.Config{
+		Host:          s.URL,
+		ContentConfig: restclient.ContentConfig{GroupVersion: &schema.GroupVersion{Group: "", Version: "v1"}},
+		Impersonate: restclient.ImpersonationConfig{
+			UserName: testUsername,
+		},
+	})
+
+	csr1.Status.Certificate = []byte("dummy data")
```

Comment:
We really need better validation on `.Status` fields.

Reply:
coming in v1